### PR TITLE
[Clang] Propagate elide safe context through [[clang::coro_await_elidable_argument]]

### PR DESCRIPTION
**Pull request by yuxuanchen1997**
Based on discussion in https://discourse.llvm.org/t/language-extension-for-better-more-deterministic-halo-for-c-coroutines/80044/9, we are proposing a new attribute `[[clang::coro_await_elidable_argument]]`. ``[[clang::coro_await_elidable_argument]]`` is a function parameter attribute. It works in conjunction with ``[[clang::coro_await_elidable]]`` to propagate a safe elide context to a parameter or parameter pack if the function is called under a safe elide context.

To give a better definition, we introduce this new concept called a "safe elide context".

When a coroutine function returns a type attributed `[[clang::coro_await_elidable]]`, a direct call expression therein that returns a prvalue of a type attributed ``[[clang::coro_await_elidable]]`` is said to be under a safe elide context if one of the following is true:
 - it is the immediate right-hand side operand to a co_await expression.
 - it is an argument to a `[[clang::coro_await_elidable_argument]]` parameter or parameter pack of another direct call expression under a safe elide context.

This attribute makes it possible to have code that composes coroutines, and the coroutines are still elidable.
```
  template <typename T>
  class [[clang::coro_await_elidable]] Task { ... };

  template <typename... T>
  class [[clang::coro_await_elidable]] WhenAll { ... };

  // `when_all` is a utility function that compose coroutines. It does not need
  // to be a coroutine to propagate.
  template <typename... T>
  WhenAll<T...> when_all([[clang::coro_await_elidable_argument]] Task<T> tasks...);

  Task<int> foo();
  Task<int> bar();
  Task<void> example1() {
    // `when_all``, `foo``, and `bar` are all elide safe because `when_all` is
    // under a safe elide context and propagated such contexts to foo and bar.
    co_await when_all(foo(), bar());
  }
```

### CODE REVIEW

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ChuanqiXu9) [inline clang/lib/Sema/SemaCoroutine.cpp:865]:
```suggestion

```

File: clang/docs/ReleaseNotes.rst
Comment (ChuanqiXu9) [inline clang/docs/ReleaseNotes.rst:254]: I don't like the name. The name were saying the entity must be awaited. But the explanation are saying other things. Let's make it more straight forward and precise. e.g, something like `[[clang::coro_must_elide_argument]]`.

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (yuxuanchen1997) [reply clang/lib/Sema/SemaCoroutine.cpp:865]: Oops.

File: clang/docs/ReleaseNotes.rst
Comment (yuxuanchen1997) [reply clang/docs/ReleaseNotes.rst:254]: Or just`[[clang::coro_await_argument]]`? If some "coro_await_elidable" is an argument that's `[[clang::coro_await_argument]]`, it indicates elidability.

Comment (ChuanqiXu9) [reply clang/docs/ReleaseNotes.rst:254]: Or elidable `[[clang::coro_elideable_await_argument]]`. I do feel better with mentioning `elide` here

Comment (vogelsgesang) [reply clang/docs/ReleaseNotes.rst:254]: What do you think about `[[clang::coro_await_elideable_argument]]`? I would slightly prefer this over `[[clang::coro_elideable_await_argument]]` because it is more similar to its sibling argument `[[clang::coro_await_elideable]]`, and the duality between the two arguments becomes more obvious already from their names.

I think we should particularly try to avoid any name with `must` in its name. To me, `must` expresses a guarantee given by the compiler to the programmer (such as for the `[[musttail]]` attribute which will fail compilation in case the call cannot be tail-called), enforced by the compiler. However, this attribute behaves the other way around: It expresses a guarantee given by the programmer to the compiler, the compiler does nothing to check or enforce the correctness of it.

**Review by vogelsgesang — COMMENTED** [review]
Thanks for implementing this! Looking forward to using it 🙂

File: clang/include/clang/Basic/AttrDocs.td
Comment (yuxuanchen1997) [reply clang/include/clang/Basic/AttrDocs.td:8266]: > If I am reading the code correctly, it is not actually a requirement that the containing / callee coroutine itself is also annotated as coro_await_elidable.

It is required. See `CurFnAwaitElidable`. Your example call to `foo()` of `NonElidableTask` won't have any elisions because we don't control `NonElidableTask::Promise::await_transform`.

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (yuxuanchen1997) [reply clang/lib/Sema/SemaCoroutine.cpp:903]: This is not the piece of code that performs this check. See `applySafeElideContext`.

But what you said is potentially a good point, we actually skipped checking two things: `operator co_await` on the callee return object, and the `await_transform`. As these operations may not be safe. The original intention was to make this a user's liability to ensure early caller destruction cannot happen in both customization points. It's not as granular as I'd like. 

With this patch, we are actually making it possible. `await_transform` and `operator co_await` here are just function calls whose arguments can be attributed with `[[clang::coro_must_await]]`.

Comment (vogelsgesang) [reply clang/lib/Sema/SemaCoroutine.cpp:903]: ah, I see. So after this patch, we could actually check `await_transform` and also `operator co_await` and thereby get rid of the `CurFnAwaitElidable` check? Such that a `ElideableTask` coawaited within a NonElidableTask (as discussed in the other thread) would become elideable?

File: clang/include/clang/Basic/AttrDocs.td
Comment (vogelsgesang) [reply clang/include/clang/Basic/AttrDocs.td:8266]: Ah, right. I misread the code.

In general: Is checking if the callee coroutine is annotated actually necessary? To my understanding from the other thread, so far it was necessary because of `await_transform` and `operator co_await`. After we make this checking more fine-granular, are there any other reasons why we would still need to check if the callee coroutine is annotated?

File: clang/docs/ReleaseNotes.rst
Comment (yuxuanchen1997) [reply clang/docs/ReleaseNotes.rst:254]: I prefer `[[clang::coro_await_elideable_argument]]`. Thanks for the ideas.

File: clang/include/clang/Basic/AttrDocs.td
Comment (yuxuanchen1997) [reply clang/include/clang/Basic/AttrDocs.td:8266]: > Is checking if the callee coroutine is annotated actually necessary?

Depends on the new spec which I haven't thought about yet but most likely I'll say it's still necessary for simplicity. We can make `await_transform` and `operator co_await` propagate the safe elide context. Just having the awaiter type attributed, and asserting that a prvalue of the awaiter returned from operator co_await must satisfy the requirements *may* do the work. I haven't thought it through.

File: clang/test/CodeGenCoroutines/coro-await-elidable.cpp
Comment (yuxuanchen1997) [reply clang/test/CodeGenCoroutines/coro-await-elidable.cpp]: Just letting you know that `$` at the end is asking the directive to literally check for a dollar sign. `{{$}}` at the end seemed to work fine.

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (vogelsgesang) [reply clang/lib/Sema/SemaCoroutine.cpp:901]: 👍

**Review by vogelsgesang — COMMENTED** [review]
Looks good to me, but at the same time: I am not experienced in the clang-frontend, so please wait for an approval of a more experienced clang-frontend contributor

**Review by ChuanqiXu9 — APPROVED** [review]

**Pull request merged**
